src/coreclr/jit/valuenum.cpp

```diff
@@ -10244,138 +10417,47 @@ bool Compiler::fgValueNumberConstLoad(GenTreeIndir* tree)
-                        if (val == 0)
-                        {
-                            tree->gtVNPair.SetBoth(vnStore->VNForNull());
-                        }
-                        else
-                        {
-                            tree->gtVNPair.SetBoth(vnStore->VNForHandle(val, GTF_ICON_OBJ_HDL));
-                            setMethodHasFrozenObjects();
-                        }
-                        return true;
-                    }
-#if defined(FEATURE_SIMD)
-                    case TYP_SIMD8:
-                    {
-                        READ_VALUE(simd8_t);
-                        tree->gtVNPair.SetBoth(vnStore->VNForSimd8Con(val));
-                        return true;
-                    }
-                    case TYP_SIMD12:
-                    {
-                        READ_VALUE(simd12_t);
-                        tree->gtVNPair.SetBoth(vnStore->VNForSimd12Con(val));
-                        return true;
-                    }
-                    case TYP_SIMD16:
-                    {
-                        READ_VALUE(simd16_t);
-                        tree->gtVNPair.SetBoth(vnStore->VNForSimd16Con(val));
-                        return true;
-                    }
-#if defined(TARGET_XARCH)
-                    case TYP_SIMD32:
-                    {
-                        READ_VALUE(simd32_t);
-                        tree->gtVNPair.SetBoth(vnStore->VNForSimd32Con(val));
-                        return true;
-                    }
-                    case TYP_SIMD64:
-                    {
-                        READ_VALUE(simd64_t);
-                        tree->gtVNPair.SetBoth(vnStore->VNForSimd64Con(val));
-                        return true;
-                    }
-#endif // TARGET_XARCH
-#endif // FEATURE_SIMD
-                    default:
-                        assert(!varTypeIsSIMD(tree));
-                        break;
+                    setMethodHasFrozenObjects();
+                }
+                tree->gtVNPair.SetBoth(vn);
+                return true;
+            }
+        }
+    }
+    else if ((varTypeIsSIMD(tree) || varTypeIsIntegral(tree) || varTypeIsFloating(tree)) &&
+             GetObjectHandleAndOffset(vnStore, tree->gtGetOp1(), &byteOffset, &obj))
+    {
+        // See if we can fold IND(ADD(FrozenObj, CNS)) to a constant
+        assert(obj != nullptr);
+        if (info.compCompHnd->isObjectImmutable(obj))
+        {
+            if ((size > 0) && (size <= maxElementSize) && ((size_t)byteOffset < INT_MAX))
+            {
+                uint8_t buffer[maxElementSize] = {0};
+                if (info.compCompHnd->readObject(obj, (uint8_t*)&buffer, size, (int)byteOffset))
+                {
+                    ValueNum vn = vnStore->VNForGenericCon(tree->TypeGet(), buffer);
+                    assert(!vnStore->IsVNObjHandle(vn));
+                    tree->gtVNPair.SetBoth(vn);
+                    return true;
```

Comment:
`readObject` makes `getArrayOrStringLength` as well as `getStringChar` redundant.

In general, the special-cased folding for `GT_ARR_LENGTH` is out-of-sync with this code, which is odd. `ARR_LENGTH` code tries to deduce the value even out of non-frozen objects.

Reply:
Good point, I also had that in mind, in theory, more specialized helpers abstract us from knowing exact object layout on JIT side but JIT already does that

Reply:
> In general, the special-cased folding for GT_ARR_LENGTH is out-of-sync with this code, which is odd. ARR_LENGTH code tries to deduce the value even out of non-frozen objects.

I'll think how to remove `getArrayOrStringLength` as well as `getStringChar`  separately if you don't mind. Currently, `getArrayOrStringLength` is the only API that can work with object handles which are not frozen.

Reply:
Added comments